pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -0,0 +1,260 @@
+	userName := nsTmplSet.GetName()
+	labels := map[string]string{
+		"owner": userName,
+	}
+	opts := client.MatchingLabels(labels)
+	namespaces := &corev1.NamespaceList{}
+	if err := r.client.List(context.TODO(), opts, namespaces); err != nil {
+		return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err, "failed to list namespace with label owner '%s'", userName)
+	}
+
+	missingNs := findNsForProvision(namespaces.Items, nsTmplSet.Spec.Namespaces, userName)
+	if missingNs != (toolchainv1alpha1.Namespace{}) {
+		nsName := toNamespaceName(userName, missingNs.Type)
+		log.Info("provisioning namespace", "namespace", missingNs)
+		if err := r.setStatusNamespaceProvisioning(nsTmplSet, nsName); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		params := make(map[string]string)
+		params["USER_NAME"] = userName
+		err := r.applyTemplate(r.client, missingNs, params)
+		if err != nil {
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to provision namespace '%s'", nsName)
+		}
+
+		// set labels
+		namespace := &corev1.Namespace{}
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Name: nsName}, namespace); err != nil {
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to get namespace '%s'", nsName)
+		}
+		if err := controllerutil.SetControllerReference(nsTmplSet, namespace, r.scheme); err != nil {
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to set controller '%s'", nsName)
+		}
+		if namespace.Labels == nil {
+			namespace.Labels = make(map[string]string)
+		}
+		namespace.Labels["owner"] = userName
+		namespace.Labels["revision"] = missingNs.Revision
+		if err := r.client.Update(context.TODO(), namespace); err != nil {
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to update namespace '%s'", nsName)
+		}
+		log.Info("namespace provisioned", "namespace", missingNs)
+		if err := r.setStatusProvisioning(nsTmplSet); err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func findNsForProvision(namespaces []corev1.Namespace, tcNamespaces []toolchainv1alpha1.Namespace, userName string) toolchainv1alpha1.Namespace {
+	for _, tcNamespace := range tcNamespaces {
+		nsName := toNamespaceName(userName, tcNamespace.Type)
+		found := findNamespace(namespaces, nsName, tcNamespace.Revision)
+		if !found {
+			return tcNamespace
+		}
+	}
+	return toolchainv1alpha1.Namespace{}
+}
+
+func findNamespace(namespaces []corev1.Namespace, namespaceName, revision string) bool {
+	for _, ns := range namespaces {
+		if ns.GetName() == namespaceName && ns.GetLabels()["revision"] == revision {
```

Comment:
So this is not only about namespace provisioning - as you check also the revision then it supports also the updates. But that's probably fine for this part of the code.